pkg/broker/handler/handler.go

```diff
@@ -45,62 +47,108 @@ type Handler struct {
 	ctx, h.cancel = context.WithCancel(ctx)
+	h.done = ctx.Done()
 	go func() {
 		defer h.cancel()
-		done(h.PubsubEvents.OpenInbound(ctx))
+		err := h.PubsubEvents.OpenInbound(ctx)
+		logging.FromContext(ctx).Error("pubsub client closed", zap.Error(err))
 	}()
 
-	curr := h.Concurrency
-	if curr <= 0 {
-		curr = 1
-	}
-	for i := 0; i < curr; i++ {
-		go h.handle(ctx)
-	}
+	go h.handle(ctx, config)
 }
 
 // Stop stops the handlers.
 func (h *Handler) Stop() {
 	h.cancel()
 }
 
-func (h *Handler) handle(ctx context.Context) {
-	for {
-		// TODO(yolocs): we need to update dep for fix:
-		// https://github.com/cloudevents/sdk-go/pull/430
-		msg, err := h.PubsubEvents.Receive(ctx)
-		if err != nil {
-			logging.FromContext(ctx).Error("failed to receive the next message from Pubsub", zap.Error(err))
-			continue
-		}
+// Done returns a channel indicating when the handler has stopped.
+func (h *Handler) Done() <-chan struct{} {
+	return h.done
+}
 
-		event, err := binding.ToEvent(ctx, msg)
-		if err != nil {
-			logging.FromContext(ctx).Error("failed to convert received message to an event", zap.Any("message", msg), zap.Error(err))
-			continue
-		}
+type pubsubResponse struct {
+	msg binding.Message
+	err error
+}
 
-		pctx := ctx
-		if h.Timeout != 0 {
-			var cancel context.CancelFunc
-			pctx, cancel = context.WithTimeout(ctx, h.Timeout)
-			defer cancel()
-		}
-		perr := h.Processor.Process(pctx, event)
-		if perr != nil {
-			logging.FromContext(pctx).Error("failed to process event", zap.Any("event", event), zap.Error(perr))
+func (h *Handler) handle(ctx context.Context, config *config.Broker) {
+	limit := h.Concurrency
+	if limit < 1 {
+		limit = 1
+	}
+	tokens := make(chan struct{}, limit)
+
+	var msgCh chan pubsubResponse
+
+	for {
+		var getToken chan struct{}
+		if msgCh == nil {
+			getToken = tokens
 		}
-		// This will ack/nack the message.
-		if err := msg.Finish(perr); err != nil {
-			logging.FromContext(ctx).Warn("failed to finish the message", zap.Any("message", msg), zap.Error(err))
+		select {
+		case getToken <- struct{}{}:
+			msgCh = make(chan pubsubResponse)
+			go func(ctx context.Context, msgCh chan<- pubsubResponse) {
+				// TODO(yolocs): we need to update dep for fix:
+				// https://github.com/cloudevents/sdk-go/pull/430
+				msg, err := h.PubsubEvents.Receive(ctx)
+				select {
+				case msgCh <- pubsubResponse{msg, err}:
+				case <-ctx.Done():
+					if err == nil {
+						if err := msg.Finish(ctx.Err()); err != nil {
+							logging.FromContext(ctx).Warn("failed to finish the message", zap.Any("message", msg), zap.Error(err))
+						}
+					}
+				}
+			}(ctx, msgCh)
+		case r := <-msgCh:
+			msgCh = nil
+			if r.err != nil {
+				logging.FromContext(ctx).Error("failed to receive the next message from Pubsub", zap.Error(r.err))
+				<-tokens
+			} else {
+				go h.handleMessage(handlerctx.WithBroker(ctx, config), r.msg, tokens)
+			}
+		case config = <-h.ConfigCh:
```

Comment:
I think it is a matter of getting used to the message passing style of communication. Maybe it would be useful to have a wider discussion on whether we should prefer this style of concurrency in general.

Reply:
My point was not about whether people can read this style. It was about whether it's necessary. Like I mentioned in the other comment, I see handler as a config-reader and it doesn't have to act on config updates, which also happens to make code simpler.

I have @ the folks who are actively working on broker to share their opinions.